README.md
# geoEventLocate
Early version of the Avenue Events project 


changelog.txt
Shit that we need to fix / add / test

Add:

Description for events
Tags for each artist to events

change main event support from lastfm to songkick
add lastfm support if songkick returns no results

Fix:

possible errors at nathans work. need to sort

Test:

Testing needs to be done on the songkick script.




Sorted (dont delete stuff just move it here)

23-11-2012 15:50 - Changelog created.

23-11-2012 22:01 redirect when at a venue and the event is happening. We know it works when the event already exists, but testing is needed for when the event is being pulled from Last.fm?

24-11-2012 20:50 Added support to use songkick to populate events.

lastFM.php
<?php

        /* 

        Loop through the massive list of LASTFM php files, cause there are many and its easier.

        */


          $files =  __DIR__ . "/LASTFM/*.php";

          foreach(glob($files) as $file) {

            //file url
            $link = $file;

            //file path info
            $info = pathinfo($link);
            //get the file name
            $filename = basename($link, '.'.$info['extension']);

            require $link;
            //require $link.$filename.".php";
          }

        /* loop to php files from specified area */

          $files1 =  __DIR__ . "/LASTFM/cache/*.php";

          foreach(glob($files1) as $file) {

            //file url
            $link = $file;

            //file path info
            $info = pathinfo($link);
            //get the file name
            $filename = basename($link, '.'.$info['extension']);

            require $link;
            //require $link.$filename.".php";
          }

          $files2 =  __DIR__ . "/LASTFM/caller/*.php";

          foreach(glob($files2) as $file) {

            //file url
            $link = $file;

            //file path info
            $info = pathinfo($link);
            //get the file name
            $filename = basename($link, '.'.$info['extension']);

            require $link;
            //require $link.$filename.".php";
          }

 
      // set api key
      CallerFactory::getDefaultCaller()->setApiKey('72cd96125291435b44055e27089f4380');

    

      //Using Last.fm to look up the Venue ID.
      $limit = 1;
      $results = Venue::search($locationName, $limit);

     
      /* Just incase we want to see whats going on.
      print "<pre>";
      print_r($results);
      print "</pre>";
      */

      //echo '<br>';
      while ($venue = $results->current()) {
      //echo "<strong>Venue ID: </strong>" . $venue->getId();


     //Now lets try and use that ID

      $eventResults = Venue::getEvents($venue->getId());

      /* Just incase we want to see whats going on.
      print "<pre>";
      print_r($eventResults[0]);
      print "</pre>";
      */



if (!empty($eventResults)) { 

         //echo ' ';

        $LASTFMartists =  $eventResults[0]->getArtists();
        
         //echo '<strong>Artists: </strong>' . $LASTFMartists[0] . "<br>";   

         $LASTFMeventTitle = $eventResults[0]->getTitle();

         //echo '<strong>Title: </strong>' . $LASTFMeventTitle . "<br>";   
      
         $LASTFMeventStartDate = $eventResults[0]->getStartDate(); 

        //echo '<strong>Start Time: </strong>' . date('d-m-Y H:i:s', $LASTFMeventStartDate);


        if (date('s', $LASTFMeventStartDate) == '01') { 
        
          $LASTFMeventTime = "19:00:00";

        } else {
        
          $LASTFMeventTime = date('H:i:s', $LASTFMeventStartDate);

      }

      
      $EM_Event = new EM_Event();

      $EM_Event->event_name = $LASTFMeventTitle;

      $EM_Event->event_start_date = date('Y-m-d', $LASTFMeventStartDate);

      $EM_Event->event_end_date = date('Y-m-d', $LASTFMeventStartDate);

      $EM_Event->event_start_time = $LASTFMeventTime;

      $EM_Event->event_end_time = "23:59:00";

      $EM_Event->event_all_day = 0;

      $EM_Event->event_owner = 1;

      $EM_Event->event_status = 1;

      $EM_Event->event_rsvp = 0;

      $EM_Event->location_id = $locationId;

      $EM_Event->group_id = 0;

      $EM_Event->start = strtotime($EM_Event->event_start_date." ".$EM_Event->event_start_time);

      $EM_Event->end = strtotime($EM_Event->event_end_date." ".$EM_Event->event_end_time);

      $EM_Event->save();

      

      if ( $locationDistance > '0.05' ) { 
      echo ' ';

      }

      require 'geoPost.php';
      
        
      }
       $venue = $results->next();

      }

      /*
      Right you little fuck face, here is the information that i can see that is in the database to create an event.

      WE HAVE TO LOOK THROUGH THE EVENTS MANAGER FILES AS THEY MUST HAVE SOME CLASSES WE CAN CALL TO LIGHTEN THE WORKLOAD!!!
      
      wp_em_events
      There are loads of columsn in this one but nearly all of them have a default value of NULL asigned and we wont need to change that. We do need the following:

      post_id - each event has to be entered into the post table and the events table. We will need to do the post table information so that we can grab the post_id and link it here.

      event slug - use the last.fm title info and replace spaces with ' - '

      event_owner - set to 1, i would guess thats admin and its the current owner of the other events. 

      event status - set to 1, i think that means its active and not a draft.

      event_name - last.fm title information

      event_end_time - set to 00:00:00 as the last.fm API does not return an end time.

      event_start_time - last.fm event start time

      event_all_day - set to 0 (1 = all day)

      event_start_date - need to split last.fm events start time variable as it contains both date and time.

      event_end_date - same as start date. Need to investigate this further in respect to festivals.

      event_rsvp_time - although we dont want to enable RSVP (the value is set to 0 by default) it looks like we have to put in a time. not sure if 00:00:00 is appropriate or the event start time. 

      location_id - use the location id from the geolocation results.

      event_atributes - not sure what this is, so i hope this makes sense to you, the current events are set to 'a:0:{}' (excluding quotes)

      event_id - set to autonumber so it should create a unique value


      wp_posts
      THIS IS A WORDPRESS TABLE - LETS NOT FUCK THIS BIT UP.

      ID - this is an autonumber but we are going to need to grab this for use in the wp_em_events table. see above.

      post_author - 1

      post_date - $theDate variable containing both date and time.

      post_date_gmt - same as post_date

      post_title - last.fm title information

      post_status - set to publish

      comment status - set to open

      ping status - set to open

      post_name - slug format of last.fm title information

      post_modified - $theDate variable

      post_modified_gmt - same if post_modified

      post_parent - 0

      guid - 'http://bethehype.co.uk/?post_type=event&#038;p=' . $ID (i hope that makes sense)

      post_type - event

      
      wp_postmeta
      GONNA TACKLE THIS BADBOY TOMORROW AS ITS A BITCH AND I AM FALLING ASLEEP.

      */

      //lets play
?>

geoPost.php
<?php

ob_start(); //remove session error

//AJAX request coordinates from HTML5 GeoLocation
$lat = $_POST['latitude'];
$lon = $_POST['longitude'];

//Write The Co-ordinates down
//**echo " Your coordinates are: " . $lat . " " . $lon;

global $currentDate, $currentTime;

//define $wpdb as a global variable
global $wpdb, $tableprefix;

if(!isset($wpdb))
{
    require_once('../../../wp-config.php');
    require_once('../../../wp-includes/wp-db.php');
}

//show any sql errors generated from the below sql query
$wpdb->show_errors();

  //select the location with the closest distance to a set of predefined coordinates then print out the details
    $location = $wpdb->get_results("SELECT location_id, location_name, location_slug, location_town, location_address, location_postcode, ((ACOS(SIN($lat * PI() / 180) * SIN(`location_latitude` * PI() / 180) + COS($lat * PI() / 180) * 
				COS(`location_latitude` * PI() / 180) * COS(($lon - `location_longitude`) * PI() / 180)) * 180 / PI()) * 60 * 1.1515) 
				AS distance FROM `wp_em_locations` HAVING distance<='20' ORDER BY distance ASC");

    //print the array(if needed)
    //print_r($result);

    //assign responsible variable names to separate parts of the array
    $locationId = $location[0]->location_id;
    $locationName = $location[0]->location_name;
    $locationTown = $location[0]->location_town;
    $locationAddress = $location[0]->location_address;
    $locationPostcode = $location[0]->location_postcode;
    $locationDistance = $location[0]->distance;

    $locationPermalink = $location[0]->location_slug; //location permalink

    //$result is an array of data in the database, print out the values of the array below

    //select the event which is showing at the above location

    $event = $wpdb->get_results("SELECT post_id, event_name, event_slug FROM wp_em_events WHERE location_id = $locationId");

    $event_post_id = $event[0]->post_id;
    $eventName = $event[0]->event_name;
    $eventPermalink = $event[0]->event_slug; //event permalink

    //separate event date notify of the status of the event

    $event_date = $wpdb->get_results("SELECT post_id, event_start_date, event_end_date, event_end_time, event_start_time FROM wp_em_events WHERE location_id = $locationId");

    //foreach ($event_date as $eventDate)

    $eventStartDate = $event_date[0]->event_start_date;
    $eventEndDate = $event_date[0]->event_end_date;
    $eventStartTime = $event_date[0]->event_start_time;
    $eventEndTime = $event_date[0]->event_end_time;

    $currentDate = date('Y-m-d');
    $currentTime = date('H:i:s');

    $theDate = $currentDate . ' ' . $currentTime;
    $beginEvent = $eventStartDate . ' ' . $eventStartTime;
    $finishEvent = $eventEndDate . ' ' . $eventEndTime;

  if ( $locationDistance < '0.05' && $beginEvent === ' ' ) {  // Within a short distance of a venue, but no events are listed

        require 'lastFM.php';

        } else if ( $locationDistance < '0.05' && $beginEvent >= $theDate ) { // Within a short distance of a venue, but no events currently or past events.
      echo ' ' . '<br>';
      echo '<h1> Okay, so you are here: </h1>';
      echo '<strong>Name:</strong> ' . $locationName;
      echo '<br>';
      echo '<strong>Town:</strong> ' . $locationTown;
      echo '<br>';
      echo '<strong>Address:</strong> ' . $locationAddress;
      echo '<br>';
      echo '<strong>PostCode:</strong> ' . $locationPostcode;
      echo '<br>';
      echo 'You are: ' . (round($locationDistance, 2)) . ' <strong>km</strong> away';
      echo '<br>';
      echo '<strong> The next event that is happening here is: </strong> ' . ' ' . $eventName . ' on ' . $eventStartDate;
      echo '<br>';
      echo '<strong> Starting at: </strong>' . $eventStartTime;

        } else if ( $locationDistance < '0.5' && $beginEvent <= $theDate && $theDate <= $finishEvent ) { // Within a short distance of a venue, and an event is currently occuring.
      
      echo $eventPermalink;


      } else if ( $locationDistance > '0.05' && $beginEvent != ' ' ) {  // Not within a short distance of a venue.
      echo ' ' . '<br>';
      echo '<h1> Your Nearest Location is </h1>';
      echo '<strong>Name:</strong> ' . $locationName;
      echo '<br>';
      echo '<strong>Town:</strong> ' . $locationTown;
      echo '<br>';
      echo '<strong>Address:</strong> ' . $locationAddress;
      echo '<br>';
      echo '<strong>PostCode:</strong> ' . $locationPostcode;
      echo '<br>';
      echo 'You are: ' . (round($locationDistance, 2)) . ' <strong>km</strong> away';
      echo '<br>';
      echo '<strong> The next event that is happening here is: </strong> ' . ' ' . $eventName;
      echo '<br>';
      echo '<strong> On: </strong>' . ' ' . $eventStartDate;
      echo '<br>';
      echo '<strong> Starting at: </strong>' . ' ' . $eventStartTime;


        } else if ( $locationDistance > '0.05' ) {  // Not within a short distance of a venue.

          require 'lastFM.php';

        } else {

      echo ' ' . '<br>';
      echo '<strong> Sorry! We havent quite worked out that situation yet!</strong>';
    }






//Shit that is commented out. Needs to be moderated!!!

    //**  echo '<br>';
    //**  echo '<strong> The Event is now happening</strong>';
    //**  echo '<br>';
      //$queried_post = get_post($event_post_id);
      //$title = $queried_post->post_title;
      //echo $title;
      //echo $queried_post->post_content;

    //** echo '<br>';
    //** echo 'Event Date: ' . $eventStartDate;
    //** echo '<br>';
    //** echo 'Event Date: ' . $eventEndDate;
    //** echo '<br>';
    //** echo 'Todays Date: ' . $currentDate;
    //** echo '<br>';

  //** echo '<br>';
  //** echo '<br>';
  //** echo '<strong>Latest Event - </strong>' . $eventName;
  //** echo '<br>';

   //if the person is within a 100m distance of the location the print a message saying you are at the location
   /*
   if ($locationDistance > '10.0') {
      echo 'No venues found.';
     } else if ($locationDistance < '0.1') {
      echo '<br>';
      echo '<br>';
      echo '<h1> You Are At: </h1>';
      echo '<br>';
      echo 'URL: ' . $locationPermalink;
      echo '<br>';
      echo '<strong>Name:</strong> ' . $locationName;
      echo '<br>';
      echo '<strong>Town:</strong> ' . $locationTown;
      echo '<br>';
      echo '<strong>Address:</strong> ' . $locationAddress;
      echo '<br>';
      echo '<strong>PostCode:</strong> ' . $locationPostcode;
      echo '<br>';
      echo 'You are: ' . (round($locationDistance, 2)) . ' <strong>km</strong> away';

   } else if ($locationDistance > '0.1') { //if the person is further than 100m distance of the location the print a message listing the nearest location
      echo '<br>';
      echo '<br>';
      echo '<h1> Your Nearest Location is </h1>';
      echo '<br>';
      echo 'URL: ' . $locationPermalink;
      echo '<br>';
      echo '<strong>Name:</strong> ' . $locationName;
      echo '<br>';
      echo '<strong>Town:</strong> ' . $locationTown;
      echo '<br>';
      echo '<strong>Address:</strong> ' . $locationAddress;
      echo '<br>';
      echo '<strong>PostCode:</strong> ' . $locationPostcode;
      echo '<br>';
      echo 'You are: ' . (round($locationDistance, 2)) . ' <strong>km</strong> away';

   }
*/
?>

songkick.php
<?php

//Attempting to use songkick instead of Last.FM

//First lets define some variables use

//Adding a new line of text to save for subversion

$SKapi ;

$SKvenueId ;

$SKvenueSearchName ;

$SKvenueSearchURL ; 

$SKvenueCalendarURL ;

// Now lets add any iformation we already know

$SKapi = 'io09K9l3ebJxmxe2' ;

$SKvenueSearchName = str_replace(' ', '%20', $locationName) ;

$SKvenueSearchURL = 'http://api.songkick.com/api/3.0/search/venues.xml?query=' . $SKvenueSearchName . '&apikey=' . $SKapi ;


// Add the wp_http wordpress class and do the request

if( !class_exists( 'WP_Http' ) )
    include_once( ABSPATH . WPINC. '../../../wp-includes/class-http.php' );

$request = new WP_Http;
$SKVenueSearchResponse = $request->request( $SKvenueSearchURL );
$SKVenueSearchResults = $SKVenueSearchResponse['body'];
$SKVenueSearchResults = simplexml_load_string($SKVenueSearchResults) ;

	 
	  /* Just incase we want to see whats going on
	  echo ' ' ;
	  print '<pre>' ;
      print_r ($SKVenueSearchResults->results->venue[0]->attributes()) ;
      print '</pre>' ;
      */

      $SKvenueId = $SKVenueSearchResults->results->venue[0]->attributes()->id ;

      
      echo ' ' ;
      // echo 'venue id: ' . $SKvenueId ;

	$SKvenueCalendarURL = 'http://api.songkick.com/api/3.0/venues/' . $SKvenueId . '/calendar.xml?apikey=' . $SKapi ;

	$request = new WP_Http;
	$SKVenueCalendarResponse = $request->request( $SKvenueCalendarURL );
	$SKVenueCalendarResults = $SKVenueCalendarResponse['body'];
	$SKVenueCalendarResults = simplexml_load_string($SKVenueCalendarResults) ;

  	  /* Just incase we want to see whats going on
	  print '<pre>' ;
      print_r ($SKVenueCalendarResults->results->event[0]) ;
      print '</pre>' ;
       */

    $SKEventTitle = $SKVenueCalendarResults->results->event[0]->attributes()->displayName ;
    $SKEventStartDate = $SKVenueCalendarResults->results->event[0]->start->attributes()->date;
    $SkEventEndDate = $SKVenueCalendarResults->results->event[0]->start->attributes()->date;
    $SKEventStartTime = $SKVenueCalendarResults->results->event[0]->start->attributes()->time;
    $SKEventHeadliner = $SKVenueCalendarResults->results->event[0]->performance->attributes()->displayName;

    if ($SKEventStartTime == '') { 
        
          $SKEventStartTime = "19:00:00";

    } 

/*
    echo '<br>' ;
    echo 'event title: ' . $SKEventTitle ;
    echo '<br>' ;
    echo 'event start date: ' . (string)$SKEventStartDate ;
    echo '<br>' ;
    echo 'headliner: ' . $SKEventHeadliner ;
	echo '<br>' ;
    echo 'event start time: ' . $SKEventStartTime ;

*/

      $EM_Event = new EM_Event();

      $EM_Event->event_name = $SKEventTitle;

      $EM_Event->event_start_date = (string)$SKEventStartDate;

      $EM_Event->event_end_date = (string)$SKEventStartDate;

      $EM_Event->event_start_time = $SKEventStartTime;

      $EM_Event->event_end_time = "23:59:00";

      $EM_Event->event_all_day = 0;

      $EM_Event->event_owner = 1;

      $EM_Event->event_status = 1;

      $EM_Event->event_rsvp = 0;

      $EM_Event->location_id = $locationId;

      $EM_Event->group_id = 0;

      $EM_Event->start = strtotime($EM_Event->event_start_date." ".$EM_Event->event_start_time);

      $EM_Event->end = strtotime($EM_Event->event_end_date." ".$EM_Event->event_end_time);

      $EM_Event->save();

      if ( $locationDistance > '0.05' ) { 
      echo ' ';

      }

      require 'geoPost.php';
      



?>

LASTFM/Album.php
<?php

//include('Media.php');

/** Represents an album and provides different methods to query album information.
 *
 * @package	php-lastfm-api
 * @author  Felix Bruns <felixbruns@web.de>
 * @version	1.0
 */
class Album extends Media {
	/** Artist of this album.
	 *
	 * @var		mixed
	 * @access	private
	 */
	private $artist;

	/** Album ID.
	 *
	 * @var		integer
	 * @access	private
	 */
	private $id;

	/** Album release date.
	 *
	 * @var		integer
	 * @access	private
	 */
	private $releaseDate;

	/** Album top tags.
	 *
	 * @var		array
	 * @access	private
	 */
	private $topTags;

	/** Create an album object.
	 *
	 * @param mixed		$artist			An artist object or string.
	 * @param string	$name			Name of this album.
	 * @param integer	$id				ID of this album.
	 * @param string	$mbid			MusicBrainz ID of this album.
	 * @param string	$url			Last.fm URL of this album.
	 * @param array		$images			An array of cover art images of different sizes.
	 * @param integer	$listeners		Number of listeners of this album.
	 * @param integer	$playCount		Play count of this album.
	 * @param integer	$releaseDate	Release date of this album.
	 * @param array		$topTags		An array of top tags of this album.
	 *
	 * @access	public
	 */
	public function __construct($artist, $name, $id, $mbid, $url, array $images,
								$listeners, $playCount, $releaseDate,
								array $topTags){
		parent::__construct($name, $mbid, $url, $images, $listeners, $playCount);

		$this->artist      = $artist;
		$this->id          = $id;
		$this->releaseDate = $releaseDate;
		$this->topTags     = $topTags;
	}

	/** Returns the artist of this album.
	 *
	 * @return	mixed	An {@link de.felixbruns.lastfm.Artist Artist} object or the artists name.
	 * @access	public
	 * @see		Artist
	 */
	public function getArtist(){
		return $this->artist;
	}

	/** Returns the ID of this album.
	 *
	 * @return	integer	The ID of this album.
	 * @access	public
	 */
	public function getId(){
		return $this->id;
	}

	/** Returns the release date of this album.
	 *
	 * @return	integer	Release date of this album as a unix timestamp.
	 * @access	public
	 */
	public function getReleaseDate(){
		return $this->releasedate;
	}

	/** Returns the top tags of this album.
	 *
	 * @return	array	An array of {@link de.felixbruns.lastfm.Tag Tag} objects.
	 * @access	public
	 * @see		Tag
	 */
	public function getTopTags(){
		return $this->topTags;
	}

	/** Tag an album using a list of user supplied tags.
	 *
	 * @param	string	$artist		The artist name in question. (Required)
	 * @param	string	$album		The album name in question. (Required)
	 * @param	array	$tags		An array of user supplied tags to apply to this album. Accepts a maximum of 10 tags. (Required)
	 * @param	Session	$session	A session obtained by {@link de.felixbruns.lastfm.Auth#getSession Auth::getSession} or {@link de.felixbruns.lastfm.Auth#getMobileSession Auth::getMobileSession}. (Required)
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function addTags($artist, $album, array $tags,
								   Session $session){
		CallerFactory::getDefaultCaller()->signedCall('album.addTags', array(
			'artist' => $artist,
			'album'  => $album,
			'tags'   => implode(',', $tags)
		), $session, 'POST');
	}

	/** Get the metadata for an album on last.fm using the album name or a MusicBrainz ID. See playlist.fetch on how to get the album playlist.
	 *
	 * @param	string	$artist	The artist name in question. (Optional)
	 * @param	string	$album	The album name in question. (Optional)
	 * @param	string	$mbid	The MusicBrainz ID for the album. (Optional)
	 * @param	string	$lang	The language to return the biography in, expressed as an ISO 639 alpha-2 code. (Optional)
	 * @return	Album			An Album object.
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function getInfo($artist, $album, $mbid = null, $lang = null){
		$xml = CallerFactory::getDefaultCaller()->call('album.getInfo', array(
			'artist' => $artist,
			'album'  => $album,
			'mbid'   => $mbid,
			'lang'   => $lang
		));

		return Album::fromSimpleXMLElement($xml);
	}

	/** Get the tags applied by an individual user to an album on last.fm.
	 *
	 * @param	string	$artist		The artist name in question. (Required)
	 * @param	string	$album		The album name in question. (Required)
	 * @param	Session	$session	A session obtained by {@link de.felixbruns.lastfm.Auth#getSession Auth::getSession} or {@link de.felixbruns.lastfm.Auth#getMobileSession Auth::getMobileSession}. (Required)
	 * @return	array				An array of Tag objects.
	 * @see		Tag
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function getTags($artist, $album, Session $session){
		$xml = CallerFactory::getDefaultCaller()->signedCall('album.getTags', array(
			'artist' => $artist,
			'album'  => $album
		), $session);

		$tags = array();

		foreach($xml->children() as $tag){
			$tags[] = Tag::fromSimpleXMLElement($tag);
		}

		return $tags;
	}

	/** Remove a user's tag from an album.
	 *
	 * @param	string	$artist		The artist name in question. (Required)
	 * @param	string	$album		The album name in question. (Required)
	 * @param	string	$tag		A single user tag to remove from this album. (Required)
	 * @param	Session	$session	A session obtained by {@link de.felixbruns.lastfm.Auth#getSession Auth::getSession} or {@link de.felixbruns.lastfm.Auth#getMobileSession Auth::getMobileSession}. (Required)
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function removeTag($artist, $album, $tag, Session $session){
		CallerFactory::getDefaultCaller()->signedCall('album.removeTag', array(
			'artist' => $artist,
			'album'  => $album,
			'tag'    => $tag
		), $session, 'POST');
	}

	/** Search for an album by name. Returns album matches sorted by relevance.
	 *
	 * @param	string	$album	The album name in question. (Required)
	 * @param	integer	$limit	Limit the number of albums returned at one time. Default (maximum) is 30. (Optional)
	 * @param	integer	$page	Scan into the results by specifying a page number. Defaults to first page. (Optional)
	 * @return	PaginatedResult	A PaginatedResult object.
	 * @see		PaginatedResult
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function search($album, $limit = null, $page = null){
		$xml = CallerFactory::getDefaultCaller()->call('album.search', array(
			'album' => $album,
			'limit' => $limit,
			'page'  => $page
		));

		$albums = array();

		foreach($xml->albummatches->children() as $album){
			$artists[] = Album::fromSimpleXMLElement($album);
		}

		$opensearch = $xml->children('http://a9.com/-/spec/opensearch/1.1/');

		return new PaginatedResult(
			Util::toInteger($opensearch->totalResults),
			Util::toInteger($opensearch->startIndex),
			Util::toInteger($opensearch->itemsPerPage),
			$artists
		);
	}

	/** Get an album playlist for streaming. INOFFICIAL.
	 *
	 * @param	string	$artist	The artist name in question. (Required)
	 * @param	string	$album	The album name in question. (Required)
	 * @return	Playlist		A Playlist object.
	 * @see		Playlist
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function getPlaylist($artist, $album){
		$xml = CallerFactory::getDefaultCaller()->call('album.getPlayerMenu', array(
			'artist' => $artist,
			'album'  => $album
		));

		return Playlist::fetch(Util::toString($xml->playlist->url), true, true);
	}

	/** Create an Album object from a SimpleXMLElement object.
	 *
	 * @param	SimpleXMLElement	$xml	A SimpleXMLElement object.
	 * @return	Album						An Album object.
	 *
	 * @static
	 * @access	public
	 * @internal
	 */
	public static function fromSimpleXMLElement(SimpleXMLElement $xml){
		$images  = array();
		$topTags = array();

		/* TODO: tagcount | library.getAlbums */

		if($xml->mbid){
			$mbid = Util::toString($xml->mbid);
		}
		else if($xml['mbid']){
			$mbid = Util::toString($xml['mbid']);
		}
		else{
			$mbid = '';
		}

		foreach($xml->image as $image){
			$images[Util::toImageType($image['size'])] = Util::toString($image);
		}

		if($xml->toptags){
			foreach($xml->toptags->children() as $tag){
				$topTags[] = Tag::fromSimpleXMLElement($tag);
			}
		}

		if($xml->artist->name && $xml->artist->mbid && $xml->artist->url){
			$artist = new Artist(
				Util::toString($xml->artist->name),
				Util::toString($xml->artist->mbid),
				Util::toString($xml->artist->url),
				array(), 0, 0, 0, array(), array(), '', 0.0
			);
		}
		else if($xml->artist && $xml->artist['mbid']){
			$artist = new Artist(
				Util::toString($xml->artist),
				Util::toString($xml->artist['mbid']),
				'', array(), 0, 0, 0, array(), array(), '', 0.0
			);
		}
		else{
			$artist = Util::toString($xml->artist);
		}

		return new Album(
			$artist,
			Util::toString($xml->name),
			Util::toInteger($xml->id),
			$mbid,
			Util::toString($xml->url),
			$images,
			Util::toInteger($xml->listeners),
			Util::toInteger($xml->playcount),
			Util::toTimestamp($xml->releasedate),
			$topTags
		);
	}
}




geoEventLocate.js
/*****

	External JavaScript file including GeoLocation and Google Map Integration

****/

//var id = document.getElementByID("content2");
function geo_location() {
if (navigator.geolocation) {
		// Use method getCurrentPosition to get coordinates
		navigator.geolocation.getCurrentPosition(function (position) {
			// Access them accordingly
			//document.getElementByID('content').innerHTML=('<p>' . position.coords.latitude . '</p>');
			//(position.coords.longitude);
			var longitude = position.coords.longitude;

			var latitude = position.coords.latitude;

			//document.getElementById('coords').innerHTML += "<input type='hidden' id='latitude' name='latitude' value=" + latitude + " /><br />";
			//document.getElementById('coords').innerHTML += "<input type='hidden' id='longitude' name='longitude' value=" + longitude + " /><br />";

			 jQuery.ajax({
       		 type:"POST",
       		 url: "/wp-content/plugins/geoEventLocate/geoPost.php?",  
      			data: "longitude=" + longitude + "&latitude="+ latitude,
        		success: function(data) {

        			//alert(data.substring(0, 1));

        			if (data.substring(0, 1) == ' ') {
        				document.getElementById("coords").innerHTML += data;
        			} else {
        				window.location = "http://bethehype.co.uk/event/" + data;
        				//holds all the php from geoPost
        				//holds all the php from geoPost
        			}
    				}
   				});
			});
		}			
	}

//$(document).ready(function(){
//$("button").click(function(){
  //$.ajax({url:"demo_test.txt",success:function(result){
    //$("#div1").html(result);
//});
//});

geoEventLocate.php
<?php
	//bethehype.co.uk
	//GeoEventLocate


		/*
		Plugin Name: GeoEventLocate
		Description: Locates an event based on current time and user location where ever the [geoEventLocate] shortcode is used.
		Version: 0.1
		Author: Nathan Brettell
		Author URI: http://nathanbrettell.com
		*/

	//Allows the plugin to be run via a shortcode from within a WordPress post
	function geoEventLocate_function() {

	//function getGeoLocation() {
		// Check for geolocation support
		$geoLocate = "<script type=\"text/javascript\" src=\"/wp-content/plugins/geoEventLocate/geoEventLocate.js\"></script>"
		. "<script> geo_location(); </script>";

			echo "<div id=\"coords\"></div>";
			echo $geoLocate;

			// SELECT ((ACOS(SIN($lat * PI() / 180) * SIN(lat * PI() / 180) + COS($lat * PI() / 180) * COS(lat * PI() / 180) * COS(($lon - lon) * PI() / 180)) * 180 / PI()) * 60 * 1.1515) 
			// AS `distance` FROM `wp_em_locations` HAVING `distance`<=’10′ ORDER BY `distance` ASC

			//$lon = $_GET['longitude'];
			//$lat = $_GET['latitude'];

			//echo "lat " . $lat . "lon" . $lon; 

		
		$theDate = date("d M Y");
		$theTime = date("H:i");
		echo '<br>';
		echo '<br>';
		echo 'The current date is: ' . $theDate . ' at ' . $theTime;
		

		//require_once('geoPost.php');

	}


add_shortcode( 'geoEventLocate', 'geoEventLocate_function');
	
?>

LASTFM/Venue.php
<?php

/** Represents a venue.
 *
 * @package	php-lastfm-api
 * @author  Felix Bruns <felixbruns@web.de>
 * @version	1.0
 */
class Venue {
	/** The venues name.
	 *
	 * @var string
	 * @access private
	 */
	private $name;

	/** The venues location.
	 *
	 * @var Location
	 * @access private
	 */
	private $location;

	/** The venues ID.
	 *
	 * @var string
	 * @access private
	 */
	private $id;

	/** The venues URL.
	 *
	 * @var string
	 * @access private
	 */
	private $url;

	

	/** Create a Venue object.
	 *
	 * @param string	$name		A venue name.
	 * @param Location	$location	A venue location.
	 * @param string	$url		A venue URL.
	 * @param string	$id		    A venue ID.
	 *
	 * @access public
	 */
	public function __construct($name, Location $location, $id, $url){
		$this->name     = $name;
		$this->location = $location;
		$this->id       = $id;
		$this->url      = $url;

	}

	/** Returns the venues name.
	 *
	 * @return	string	A venue name.
	 * @access public
	 */
	public function getName(){
		return $this->name;
	}

	/** Returns the venues location.
	 *
	 * @return	Location	A venue location.
	 * @access public
	 */
	public function getLocation(){
		return $this->location;
	}

	/** Returns the venues URL.
	 *
	 * @return	string	A venue URL.
	 * @access public
	 */
	public function getId(){
		return $this->id;
	}

	/** Returns the venues ID.
	 *
	 * @return	string	A venue ID.
	 * @access public
	 */
	public function getUrl(){
		return $this->url;
	}



	/** Get a list of upcoming events at this venue.
	 *
	 * @param	string	$venue	The venue id to fetch the events for. (Required)
	 * @return	array			An array of Event objects.
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function getEvents($venue){
		$xml = CallerFactory::getDefaultCaller()->call('venue.getEvents', array(
			'venue' => $venue
		));

		$events = array();

		foreach($xml->children() as $event){
			$events[] = Event::fromSimpleXMLElement($event);
		}

		return $events;
	}

	/** Get a paginated list of all the events held at this venue in the past.
	 *
	 * @param	string	$venue	The id for the venue you would like to fetch event listings for. (Required)
	 * @param	integer	$limit	The maximum number of results to return. (Optional)
	 * @param	integer	$page	The page of results to return. (Optional)
	 * @return	PaginatedResult	A PaginatedResult object.
	 * @see		PaginatedResult
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function getPastEvents($venue, $limit = null, $page = null){
		$xml = CallerFactory::getDefaultCaller()->call('venue.getPastEvents', array(
			'venue' => $venue,
			'limit' => $limit,
			'page'  => $page
		));

		$events = array();

		foreach($xml->children() as $event){
			$events[] = Event::fromSimpleXMLElement($event);
		}

		$perPage = Util::toInteger($xml['perPage']);

		return new PaginatedResult(
			Util::toInteger($xml['total']),
			(Util::toInteger($xml['page']) - 1) * $perPage,
			$perPage,
			$events
		);
	}

	/** Search for a venue by venue name .
	 *
	 * @param	string	$venue		The venue name you would like to search for. (Required)
	 * @param	integer	$limit		The number of results to fetch per page. Defaults to 50. (Optional)
	 * @param	integer	$page		The results page you would like to fetch. (Optional)
	 * @param	string	$country	Filter your results by country. Expressed as an ISO 3166-2 code. (Optional)
	 * @return	PaginatedResult		A PaginatedResult object.
	 * @see		PaginatedResult
	 *
	 * @static
	 * @access	public
	 * @throws	Error
	 */
	public static function search($venue, $limit = null, $page = null, $country = null){
		$xml = CallerFactory::getDefaultCaller()->call('venue.search', array(
			'venue'   => $venue,
			'limit'   => $limit,
			'page'    => $page,
			'country' => $country
		));

		$venues = array();

		foreach($xml->venuematches->children() as $venue){
			$venues[] = Venue::fromSimpleXMLElement($venue);
		}


		$opensearch = $xml->children('http://a9.com/-/spec/opensearch/1.1/');

		return new PaginatedResult(
			Util::toInteger($opensearch->totalResults),
			Util::toInteger($opensearch->startIndex),
			Util::toInteger($opensearch->itemsPerPage),
			$venues
		);
	}

	/** Create a Venue object from a SimpleXMLElement.
	 *
	 * @param	SimpleXMLElement	$xml	A SimpleXMLElement.
	 * @return	Venue						A Venue object.
	 *
	 * @static
	 * @access	public
	 * @internal
	 */
	public static function fromSimpleXMLElement(SimpleXMLElement $xml){
		return new Venue(
			Util::toString($xml->name),
			Location::fromSimpleXMLElement($xml->location),
			Util::toString($xml->id),
			Util::toString($xml->url)
		);
	}
}


